_drafts/heartbeats.md

```diff
@@ -0,0 +1,968 @@
+
+<hr>
+
+$$\left(1-\frac{2000}{24000}\right) \times 100 = \textbf{91.6% crash-free hours}$$
+
+This is a very good metric to use to gauge stability between releases, and it's
+incredibly simple to calculate once the data is in a database table.
+
+### Battery Life Trends
+
+If each device records the percentage or mV
+[battery drain per hour](#battery-life-drain-metric) in a heartbeat metric, then
+it can easily project what the battery life of a firmware release will be with
+only a handful of devices over just a few hours. This is _incredibly_ useful if
+your devices have a battery life of weeks and you would rather not wait weeks
+for a device to complete its charge-discharge cycle.
+
+To do this, simply record the delta in battery life during the heartbeat
+interval (assuming the device was not plugged into a charger) in a gauge metric,
+and then take the average of this delta over **all heartbeats** of the
+particular release.
+
+<div>
+<span>$$\left(100 \div \frac{total\:battery\:pct\:drained\:over\:all\:heartbeats}{\#\:of\:heartbeats}\right)$$</span>
+<span>$$ = \textbf{projected days battery life}$$</span>
+</div>
+
+### Wi-Fi / BLE Connected Hours
+
+This is a crucial metric to track for devices that connect to a plethora of
+Wi-Fi routers and mobile phones on various versions of iOS or Android. Knowing
+from past experience, it seems like _anything_ can affect connectivity
+performance and there are constantly regressions.
+
+Once again, the metric is very easily to calculate assuming that the
+[Wi-Fi / BLE connectivity is tracked](#wifi-ble-connectivity-metric) within an
+hourly heartbeat.
+
+<div>
+<span>$$\left(\frac{total\:time\:connected\:over\:all\:heartbeats}{\#\:of\:heartbeats}\right) \times 100$$</span>
+<span>$$ = \textbf{ % connected hours}$$</span>
+</div>
+
+## What's Next?
+
+In this article, I've primarily covered how to think about, track, and store
+heartbeat metrics and how they compare against logging. I have completely
+ignored how to get the metrics **off** of the device and into something that
+developers and decision-makers can query. This is on purpose.
+
+I imagine that every company has their favorite cloud vendor, their favorite
+database or data warehouse, and systems in place around device and data privacy.
+I could produce a one-off solution that a single developer could prop up
+themselves, but collecting metrics and making business decisions from these
+metrics is a problem that the entire company should be on board with.
+
+Bring in an infrastructure engineer from your company into the conversations and
+ask them how best to ingest and query these heartbeat metrics.
+
+If you'd rather just have someone else do it entirely, reach out to one of us at
+Memfault. We'd be happy to guide you in the right direction and suggest a
+solution, whether it's our solution or not.
+
+## Conclusion
+
+<!-- Interrupt Keep START -->
+
+{% include newsletter.html %}
```

Comment:
idea: maybe we should make these entire boxes a link (instead of just "Subscribe" & "Github", respectively)

Reply:
I like it!